File: src/main/java/org/example/testscala/instrumentation/cats/CatsInstrumentation.java
package org.example.testscala.instrumentation.cats;

import kanela.agent.api.instrumentation.InstrumentationBuilder;

public class CatsInstrumentation extends InstrumentationBuilder {

  public CatsInstrumentation() {
    //    onSubTypesOf("scala.concurrent.ExecutionContext").advise(
    //      method("execute").and(withArgument(Runnable.class)),
    //      CatsExecutionContextAdvisor.class
    //    );

    //    onSubTypesOf("cats.effect.unsafe.Scheduler").advise(
    //      method("sleep").and(withArgument(0, FiniteDuration.class)).and(withArgument(1, Runnable.class)),
    //      CatsSchedulerAdvisor.class
    //    );

    onType("cats.effect.IOFiber")
      .advise(
        method("runLoop").or(method("succeeded")).or(method("failed")),
        CatsFiberAdvisor.class
      );
  }

}

File: src/main/scala/instrumentation/CatsInstrumentation.scala
package org.example.testsscala
package instrumentation

import kamon.Kamon
import kamon.instrumentation.context.HasContext
import kanela.agent.api.instrumentation.InstrumentationBuilder
import kanela.agent.libs.net.bytebuddy.asm.Advice

class CatsInstrumentation extends InstrumentationBuilder {

  println(s"Called instrumentation: ${getClass}")

  onType("cats.effect.IOFiber")
    .mixin(classOf[HasContext.Mixin])
    .advise(method("run"), classOf[CatsFiberRunAdvisor])
    .advise(method("scheduleFiber").or(method("rescheduleFiber")), classOf[CatsFiberScheduleAdvisor])

}

class CatsFiberRunAdvisor
object CatsFiberRunAdvisor {

  @Advice.OnMethodEnter
  def beforeRunCall(@Advice.This fiberWithContext: HasContext): Unit = {
    val context = fiberWithContext.context
    Kamon.storeContext(context)
    // println(s"[${Thread.currentThread} -- $fiberWithContext -- Run] Before - Set current context to $context from fiber")
    // println(s"Running fiber $fiberWithContext")
  }

  @Advice.OnMethodExit
  def afterRunCall(@Advice.This fiberWithContext: HasContext): Unit = {
    val context = Kamon.currentContext()
    fiberWithContext.setContext(context)
    // println(s"[${Thread.currentThread} -- $fiberWithContext -- Run] After - Saving context $context in fiber")
  }

}

class CatsFiberScheduleAdvisor
object CatsFiberScheduleAdvisor {

  @Advice.OnMethodEnter
  def beforeScheduleCall(@Advice.This fiberWithContext: HasContext, @Advice.Argument(1) fiberWithContextArg: HasContext): Unit = {
    // val context = fiberWithContext.context
    // Kamon.storeContext(context)
    // println(s"[${Thread.currentThread} -- $fiberWithContext -- Schedule] Before")
    // println(s"Running fiber $fiberWithContext")

    if (fiberWithContext ne fiberWithContextArg) {
      // println(s"[${Thread.currentThread} -- $fiberWithContext -- Schedule] Before -- Different fibers, coping current Kamon context to the new Fiber")
      fiberWithContextArg.setContext(Kamon.currentContext())
    }
  }

  @Advice.OnMethodExit
  def afterScheduleCall(@Advice.This fiberWithContext: HasContext, @Advice.Argument(1) fiberWithContextArg: HasContext): Unit = {
    // val context = Kamon.currentContext()
    // fiberWithContext.setContext(context)
    // println(s"[${Thread.currentThread} -- $fiberWithContext -- Schedule] After")
  }

}

